### PR TITLE
Add extensionsTab to preferences page

### LINKED ISSUE #6530: Add extensions panel to About:Preferences
## Test plan
See https://github.com/brave/browser-laptop/pull/8065

## Original issue description
To increase awareness of Brave's extension support, and to funnel new extension requests, a new prefs panel will be created.

features:
- all brave extensions will appear in a sortable list, sorted alphabetically by default
- each extension can be enabled/disabled with a switch at the right side
(this should toggle the presence of the extension in the point of integration)
- if an extension is available but not installed by default, the switch begins as an "install" button
- a rt-click menu is available to uninstall any extension that was once manually installed
- a notification bar is triggered when new extensions become available to alert the user of the good news (this alert may be disabled)
- a link is positioned at the bottom of the page to invite users to submit extension requests through the community portal

notes: 
- this panel is for extensions only and will not display plugins such as Flash or Widevine
- after implementation, we will add a menu item to point users to this panel
(Window>Extensions Manager...)

pasted image at 2017_02_27 04_21 pm

### PR DESCRIPTION
**Pull request by cezaraugusto**
- [x] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- [x] Ran `git rebase -i` to squash commits (if needed).

Add extensionsTab to preferences page

Auditors: @bsclifton, @bbondy

Closes #6530

screen shot 2017-04-04 at 4 10 08 pm

## Test plan:

1. Automated test must pass:

_extension state_
```
npm run test -- --grep="extensionUninstalled"
```

_extensionsReducer_
```
npm run test -- --grep="extensionsReducer"
```

_extensionsTab_
```
npm run test -- --grep="ExtensionsTab component"
```

## QA Steps:

1. Go to extensions tab
2. Enabling an extension should enable it
3. Disabling an extension should disable it
4. Removing an extension should remove its folder (you can see its path logged on console)
5. Removed extension should not appear in UI, and should be disabled after removal
6. Enabling a removed extension should install it again
7. Enabling another password manager should disable the first
8. Disabling all password managers should set "built-in" password manager to active
9. You can't remove built-in (webtorrent and sync) extensions -- **update** per @bradleyrichter request Sync is no longer included

### CODE REVIEW

**Comment by jonathansampson** [comment]
@cezaraugusto Rebase, please? :)

**Comment by jonathansampson** [comment]
Perhaps _about:extensions_ should redirect to _about:preferences#extensions_ with this change?

**Comment by jonathansampson** [comment]
I don't feel too great about including these extensions images in the browser itself. I feel as though we should get those from the extension store instead. Otherwise, we'll have to do extra work each time we support a new extension.

**Comment by cezaraugusto** [comment]
rebased. Added more tests.

**Comment by cezaraugusto** [comment]
@jonathansampson I agree with you regarding extensions images but at this moment there's no way of having their data unless it's installed by default. 

Ideally they would have their store with name, description, and image, even when they're not installed by default (built-in). At that moment we can also remove dummyContent I included to show them all

**Comment by jonathansampson** [comment]
@cezaraugusto Good point. Can't show an extension's image if the image hasn't yet been loaded with the extension. My apologies :)

File: app/extensions/brave/locales/en-US/extensions.properties [hunk 1]
```diff
@@ -3,3 +3,29 @@ extensions=Extensions
...
+l10nOnepasswordDesc=1Password extension for Brave.
+l10nPdfjs= PDF Viewer
+l10nPdfjsDesc=Uses HTML5 to display PDF files directly in the browser.
+l10nPocket= Save to Pocket
+l10nPocketDesc=Pocket Extension for Brave - The best way to save articles, videos and more.
+l10nSync= Brave Sync
+l10nSyncDesc=
+l10nWebtorrent= Torrent Viewer
+l10nWebtorrentDesc=
+l10nVimium= Vimium
```
Comment (jonathansampson) [inline app/extensions/brave/locales/en-US/extensions.properties]: We'll be shipping Honey soon (perhaps in 0.14.2). Can we include it?
https://chrome.google.com/webstore/detail/honey/bmnlcjabgnpnenekpadlanbbkooimhnj

File: app/renderer/components/settings.js [hunk 2]
```diff
@@ -58,10 +59,20 @@ class SettingCheckbox extends ImmutableComponent {
   }
 
   onClick (e) {
+    if (this.props.forPassword) {
+      // You can only have one active password manager
+      // if user decide to disable all,
+      // switch back password to unmanaged (null)
+      e.target.value
+        ? aboutActions.changeSetting(settings.ACTIVE_PASSWORD_MANAGER, this.props.prefKey)
+        : aboutActions.changeSetting(settings.ACTIVE_PASSWORD_MANAGER, null)
```
Comment (jonathansampson) [inline app/renderer/components/settings.js]: Similar pattern in _extensionsTab.js_. There `false` was used; here `null` is being used. Can we go with one or the other, so there isn't any confusion about whether or not the two behave differently?

```
+    isPasswordManager(extensionId)
+      ? aboutActions.changeSetting(settings.ACTIVE_PASSWORD_MANAGER, extensionId)
+      : aboutActions.changeSetting(getExtensionSetting(extensionId), false)
```

File: app/renderer/lib/extensionsUtil.js [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+const brave = config.braveExtensionId
+const dashlane = extensionIds[passwordManagers.DASHLANE]
+const enpass = extensionIds[passwordManagers.ENPASS]
+const lastpass = extensionIds[passwordManagers.LAST_PASS]
+const onepassword = extensionIds[passwordManagers.ONE_PASSWORD]
+const pdfjs = config.PDFJSExtensionId
+const pocket = config.PocketExtensionId
+const sync = config.syncExtensionId
+const webtorrent = config.torrentExtensionId
+const vimium = config.vimiumExtensionId
```
Comment (jonathansampson) [inline app/renderer/lib/extensionsUtil.js]: Please add Honey :)

File: app/renderer/lib/extensionsUtil.js [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    name: 'l10nOnepassword',
+    description: 'l10nOnepasswordDesc',
+    icon: 'img/extensions/1password-128.png'
+  }, {
+    id: pocket,
+    name: 'l10nPocket',
+    description: 'l10nPocketDesc',
+    icon: 'img/extensions/pocket-128.png'
+  }
+  // { id: 'vimium' // TBD }
```
Comment (jonathansampson) [inline app/renderer/lib/extensionsUtil.js]: Please add Honey.

File: app/renderer/lib/extensionsUtil.js [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    case pocket:
+      extensionSetting = settings.POCKET_ENABLED
+      break
+    case webtorrent:
+      extensionSetting = settings.TORRENT_VIEWER_ENABLED
+      break
+    case vimium:
+      extensionSetting = settings.VIMIUM_ENABLED
+      break
+    default:
```
Comment (jonathansampson) [inline app/renderer/lib/extensionsUtil.js]: Please add Honey.

File: js/components/main.js [hunk 6]
```diff
@@ -1339,7 +1339,7 @@ class Main extends ImmutableComponent {
                         .includes(siteTags.BOOKMARK)) || emptyMap
                   : null}
                 history={this.bindHistory(frame)}
-                extensions={frame.get('location') === 'about:extensions'
+                extensions={['about:extensions', 'about:preferences'].includes(getBaseUrl(frame.get('location')))
```
Comment (jonathansampson) [inline js/components/main.js]: Should we continue to support `about:extensions` after this?

File: app/renderer/components/settings.js (hunk 2 above)
Comment (cezaraugusto) [inline app/renderer/components/settings.js]: yes I'll update, thanks for noticing

**Comment by bbondy** [comment]
@cezaraugusto pls rebase this one when you get a chance. Thanks!

**Comment by cezaraugusto** [comment]
ok rebased 👍

File: app/extensions/brave/locales/en-US/extensions.properties (hunk 1 above)
Comment (cezaraugusto) [inline app/extensions/brave/locales/en-US/extensions.properties]: not sure which is going to land first but will include once Honey lands

**Comment by luixxiul** [comment]
@cezaraugusto does this close #6794 as well?

**Comment by bsclifton** [comment]
@luixxiul it sure does! 😄  I'll go ahead and close that out

**Pull request merged**